keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,147 @@
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `pod.Spec.SecurityContext` field expanded to include a new field called `VolumePermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership. This could be useful to users when security policy of a cluster prevents users from removing `fsGroup` from their pod's `SecurityContext`.
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented. Algorithm that performs permission change however will be changed to perform permission change of top level directory last.
+ - `OnDemand` --> Only perform permission change if permissions of top level directory does not match with expected permissions.
+
+```go
+type PodVolumePermissionChangePolicy string
+
+const(
+    NeverChangeVolumePermission PodVolumePermissionChangePolicy = "NoChange"
+    OnDemandChangeVolumePermission PodVolumePermissionChangePolicy = "OnDemand"
+    AlwaysChangeVolumePermission PodVolumePermissionChangePolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // VolumePermissionChangePolicy ← new field
+    // + optional
+    VolumePermissionChangePolicy *PodVolumePermissionChangePolicy
+}
+```
+
+### Risks and Mitigations
+
+- One of the risks is if user volume's permission was previously changed using old algorithm(which changes permission of top level directory first) and user opts in for `OnDemand` `VolumePermissionChangePolicy` then we can't distinguish if the volume was previously only partially recursively chown'd.
+
+
+## Graduation Criteria
+
+* Alpha in 1.18 provided all tests are passing and gated by the feature Gate
+   ConfigurableVolumeFilePermissions and set to a default of `False`
+
+* Beta in 1.19 with design validated by at least two customer deployments
+  (non-production), with discussions in SIG-Storage regarding success of
+  deployments.  A metric will be added to report time taken to perform a
+  volume ownership change.
+* GA in 1.20, with Node E2E tests in place tagged with feature Storage
+
+
+[umbrella issues]: https://github.com/kubernetes/kubernetes/issues/69699
+
+### Test Plan
+
+A test plan will consist of the following tests
+
+* Basic tests including a permutation of the following values
+  - PodSecurityContext.VolumePermissionChangePolicy (Never/OnDemand/Always)
+  - PersistentVolumeClaimStatus.FSGroup (matching, non-matching)
+  - Volume Filesystem existing permissions (none, matching, non-matching, partial-matching?)
+* E2E tests
+
+
+### Monitoring
+
+We will add a metric that measures the volume ownership change times.
+
+## Implementation History
+
+- 2020-01-20 Initial KEP pull request submitted
+
+## Drawbacks [optional]
+
+
+## Alternatives [optional]
+
```

Comment:
Should we briefly talk about the idea of redesigning the volume permissions API?

Reply:
done. Also added bit about shiftfs